Use logr with zap throughout

rainest opened the pull request:
**What this PR does / why we need it**:

Converts pure logrus subsystems to logr.

Uses zapr+zap in place of logrusr+logrus to create the logr.Logger.

**Which issue this PR fixes**:

Makes logging consistent throughout the project, except for the odd KLog line from client-go and one hack script that I don't really care about.

Allows swapping out logging engines without touching individual log lines.

Fix #1893 

**Special notes for your reviewer**:

Review open for code. Want to discuss the below in sync before merging or building additional functionality.

- There is [no equivalent to the logrus console format available in zap](https://github.com/Kong/kubernetes-ingress-controller/issues/1893#issuecomment-1730599566) without rolling our own, though there's an open community PR we could build off. This has been left with the zap default (tab-delimited, field names omitted) with JSON as the designated "feed it into your structured log SIEM" option.
- Hopes for zapr handling verbosity to level conversions were a bit overly optimistic. The truth is [kinda weird](https://github.com/Kong/kubernetes-ingress-controller/issues/1893#issuecomment-1732021305). I've set the logger to omit the level and use the actual verbosity instead. We could reasonably re-enable level with the understanding that there's no longer `WARN`, since neither we nor controller-runtime use levels greater than 1 (debug, effectively) often, so the info -> debug -> info again progression doesn't really show up. We could try fixing that upstream.
- I added superfluous verbosity setters (`V(util.InfoLevel).Info()` and the like) to log lines I converted. These don't do anything useful at present, but I figure it lets us change them later via the constants if we decide to do something fancier with the verbosity levels.

**PR Readiness Checklist**:

Complete these before marking the PR as `ready to review`:

- [X] the `CHANGELOG.md` release notes have been updated to reflect any significant (and particularly user-facing) changes introduced by this PR

programmer04 commented on internal/dataplane/parser/translate_udproute.go:
```diff
@@ -56,7 +57,7 @@ func (p *Parser) ingressRulesFromUDPRoutes() ingressRules {
 
 	if len(errs) > 0 {
 		for _, err := range errs {
-			p.logger.Errorf(err.Error())
+			p.logger.V(util.ErrorLevel).Error(err, "could not generate route from UDPRoute")
 		}
 	}
```
Having just is enough, `if` is redundant
```go
for _, err := range errs {
   p.logger.V(util.ErrorLevel).Error(err, "could not generate route from UDPRoute")
}
```

programmer04 commented on internal/util/logging.go:
```diff
@@ -4,75 +4,105 @@ import (
-	// InfoLevel is the converted logging level from logrus to go-logr for
-	// information level logging. Note that the logrusr middleware technically
-	// flattens all levels prior to this level into this level as well.
-	InfoLevel = int(logrus.InfoLevel) - logrusrDiff
+	// InfoLevel is the logr verbosity level for info logs.
+	InfoLevel = 0
 
-	// DebugLevel is the converted logging level from logrus to go-logr for
-	// debug level logging.
-	DebugLevel = int(logrus.DebugLevel) - logrusrDiff
+	// DebugLevel is the logr verbosity level for debug logs.
+	DebugLevel = 1
 
-	// WarnLevel is the converted logrus level to go-logr for warnings.
-	WarnLevel = int(logrus.WarnLevel) - logrusrDiff
+	// TraceLevel is the logr verbosity level for trace logs.
+	TraceLevel = 2
```
Maybe it's worth adjusting the below

https://github.com/Kong/kubernetes-ingress-controller/blob/96d8ed2508ba7478ee3d20c645db091ad66c66c5/internal/manager/config.go#L144

since we're releasing the new major. Do  I understand correctly that it doesn't matter whether I choose `error`, `warn`, `info` the same stuff will be logged? Levels `fatal` and `panic` seems redundant

rainest replied:
TBD, I wanted to discuss how we actually handle these in sync, since there are a few different options. As-is the levels are squashed, but the config is unchanged to avoid breaking existing config.

programmer04 replied:
There is a dedicated issue for making such changes in KIC 3.0.0 milestone
- https://github.com/Kong/kubernetes-ingress-controller/issues/4729

programmer04 commented on internal/util/logging.go:
```diff
@@ -4,75 +4,105 @@ import (
-// our logging implementations into one or the other.
-//
-// See: https://github.com/Kong/kubernetes-ingress-controller/issues/1893
+// logr does not use semantic log types. Logs are either informational or error messages with arbitrary verbosity,
```
I'd suggest moving it out of the package `utils` to some log-dedicated package with a better name e.g. `konglog` or something like that

programmer04 replied:
Maybe in this package helpers like `NewNoOpLogger()` could be introduced and used across the codebase rather than explicit 
```go
zapr.NewLogger(zap.NewNop())
```
in many places, it encapsulates `zapr` and `zap` making them implementation detail. Maybe they should be only imported in this package... Something to consider

rainest replied:
The places where you set up a logger aren't implementation agnostic. logr defines how you log calls, but not how you manipulate the underlying logger. This mostly matters for tests (the application code should use the same base logger created by `SetupLoggers()` throughout), and the way zap and logrus handle log observation doesn't follow any standard interface or immediately lend itself to one.

If the setup code required is short, I do tend to just copy/paste it, since I'd rather that than try to expand a helper when I hit some future test that has breaking needs the original helper didn't account for.

programmer04 commented on internal/manager/setup.go:
```diff
@@ -38,25 +37,19 @@ import (
-	}
-
-	logger := logrusr.New(deprecatedLogger)
+	logger := zapr.NewLoggerWithOptions(zapBase, zapr.LogInfoLevel("v"))
```
Maybe this line should be moved to `util.MakeLogger`? I think that we want rather always include this field. WDYT?

rainest replied:
Dunno. I just kept the original layout, where there was more of a distinction. MakeLogger created the underlying logging engine and returned it to SetupLoggers, and SetupLoggers then returned both that and a logr wrapper it created. Keeping the "build a thing that can be logr-ed" and "build a logr from some logr-able thing" code separate still kinda makes sense even if we now only return one.

programmer04 reviewed (commented):
Great effort (I can't wait for the merge)🎖️ 

FYI: in `go-logr` docs, there is a good [go-logr vs. slog](https://github.com/go-logr/logr?tab=readme-ov-file#background), worth checking. `slog` is less feature reach

rainest commented on internal/dataplane/failures/failures_test.go:
```diff
@@ -61,53 +64,52 @@ func TestResourceFailure(t *testing.T) {
+		lastLog := logs.All()[logs.Len()-1]
 		require.NotNil(t, lastLog)
-		require.Equal(t, logrus.WarnLevel, lastLog.Level)
+		// TODO 1893 idk how this conversion should actually work, and fiddling with values is tiresome
+		// it will work when the level conversions are sorted. the code does log the warn constant
+		require.Equal(t, zapcore.Level(util.WarnLevel), lastLog.Level)
```
Removed; forgot to strip this out after it was un-broken.

rainest commented on internal/dataplane/kongstate/route.go:
```diff
@@ -197,15 +197,15 @@ func (r *Route) overrideMethods(log logrus.FieldLogger, anns map[string]string)
 			// if any method is invalid (not an uppercase alpha string),
 			// discard everything
-			log.WithField("kongroute", r.Name).Errorf("invalid method: %v", method)
+			logger.V(util.ErrorLevel).Error(nil, "invalid method", "route_name", r.Name, "method", method)
```
Maybe. It's ignored but the value is still available. Whether that value is printed is implementation-dependent. zapr at least [only includes it for Info](https://github.com/go-logr/zapr/blob/fe8a3ccb2b17db0a1602e5a39b6221b40a7ed4fb/zapr.go#L274-L291).

I originally included it somewhat arbitrarily to make log calls consistent, and because there used to be some `V(util.DebugLevel).Error()` calls from `logrus.Errorf("something %w", err)` calls at level debug. The latter I've since converted to `Info()` as I think we cared more about the threshold and happened to have an error from upstream that we could include.

rainest commented on internal/util/debug_logging.go:
```diff
@@ -1,176 +0,0 @@
-package util
```
I'm just immediately removing it, since it shouldn't be commonly used. If someone complains we have reason to add sampling configuration.

rainest commented:
From sync/other feedback:

- `Error()` calls no longer explicitly indicate an error V level.
- Restored the level key. We'd need to remove it or change something in libraries to use more verbosity levels, but we're going to follow controller-runtime's basic info/error/debug, where the zapr level handling is correct.
- Removed the redundant config log levels. They're just a minor breaking change now.
- Added a `warning` key to allow filtering those still.

pmalek reviewed (commented):
Given that we got rid of `logrus` completely in this PR I suggest we add a `golangci-lint` rule to prevent re-adding it in the future.

I've pushed a change that enforces that https://github.com/Kong/kubernetes-ingress-controller/pull/4688/commits/d449a3339a26c9d78691826340c2c64a58de01f9. Feel free to change it/remove it from this PR.

The only leftover transient dependency on logrus we have is coming from kubernetes-telemetry module now. We might want to reconsider our log usage there.

rainest commented:
We can't do anything here for it though. It'd need to be refactored to accept a logr.Logger from its importer rather than always creating its own. That issue sorta exists already: https://github.com/Kong/kubernetes-telemetry/issues/48#issuecomment-1741474936

pmalek commented:
> We can't do anything here for it though. It'd need to be refactored to accept a logr.Logger from its importer rather than always creating its own. That issue in [Kong/kubernetes-telemetry#48](https://github.com/Kong/kubernetes-telemetry/issues/48) and [Kong/kubernetes-telemetry#46](https://github.com/Kong/kubernetes-telemetry/issues/46).
> 
> IDK that telemetry strictly needs a default--I guess you technically can run it independently, but I don't know why anyone would. There's also a split between consumer and manager loggers that [we should probably refactor](https://github.com/Kong/kubernetes-telemetry/issues/46#issuecomment-1741515762).

Good, then perhaps it's a good time to prioritize this 👍

pmalek commented:
@rainest Can you attach pics/snippets that compare before and after?

rainest commented:
> @rainest Can you attach pics/snippets that compare before and after?

Short version:

console/text now (logrus):
```
time="2023-09-21T23:07:26Z" level=info msg="the ingress class name has been set" logger=setup value=kong othervalue=pong
```

console/text after (zap)
```
2023-09-22T00:38:16.026Z        info    setup   the ingress class name has been set     {"value": "kong","othervalue":"pong"}
```

json now (logrus):
```
{"level":"info","logger":"setup","msg":"the ingress class name has been set","time":"2023-09-21T23:15:15Z","value":"kong"}
```

json after (zap):
```
{"level":"info","time":"2023-09-22T00:28:13.006Z","logger":"setup","msg":"the ingress class name has been set","value":"kong"}
```

For more detail, https://gist.github.com/rainest/261e6075656603364854220262d6a746 is a longer example with different sorts of calls.

Those are from stuffing the example logs into manager start on this branch and main, and then running integration tests to capture it. It looks like we also had some formatting discrepancies between test and regular loggers, since [some logs I captured earlier](https://github.com/Kong/kubernetes-ingress-controller/issues/1893#issuecomment-1730599566) from a normally deployed container have logrus using `key=value` for level and timestamp also, whereas the detailed examples from the test logger don't.

One notable bit is that you can kinda use zap json as something similar to the old full `key=value` console since it add fields in a consistent order (level, timestamp, logger name, message, structured fields in the order they were added), whereas logrus alpha sorts the JSON keys (usually sticking level and timestamp somewhere in the middle).

pmalek commented on CHANGELOG.md:
```diff
@@ -85,6 +85,15 @@ Adding a new version? You'll need three changes:
   [#4748](https://github.com/Kong/kubernetes-ingress-controller/pull/4748)
+- Removed support for the `debug-log-reduce-redundancy` CLI flag.
+  [#4688](https://github.com/Kong/kubernetes-ingress-controller/pull/4688)
+- The "text" logging format has changed. "json" should be used for
+  machine-parseable logs.
+  [#4688](https://github.com/Kong/kubernetes-ingress-controller/pull/4688)
```
I can see that JSON log format changes the timestamp format, from 

    "time":"2023-09-21T23:15:15Z"

to

    "time":"2023-09-22T00:28:13.006Z"

Which is a subtle change but might break some log ingestion systems/rules/etc.

This should definitely be noted in here as for some this will be a breaking change.

rainest replied:
Good catch. Switched to RFC3339, which does ISO8601 with second precision. Console format because that's what the tests run in, but the time encoder is the same:

```
2023-10-10T00:49:13Z	info	setup	Initializing Dataplane Client	{"v": 0}
2023-10-10T00:49:13Z	info	setup	Starting Admission Server	{"v": 0}
```

pmalek commented on CHANGELOG.md:
```diff
@@ -85,6 +85,15 @@ Adding a new version? You'll need three changes:
+  machine-parseable logs.
+  [#4688](https://github.com/Kong/kubernetes-ingress-controller/pull/4688)
+- The "warn", "fatal", and "panic" log levels are no longer available. "error"
+  is now the highest log level. "warn" logs are now logged at "info" level, but
+  include an additional "warning=true" field.
```
> "warn" logs are now logged at "info" level, but include an additional "warning=true" field

What's the mechanism for that? 🤔 Could we "support" the warning log level? For example by introducing our own logger interface that embeds the `logr.Logger` and adds a warning level (and appropriate method)? Obviously we'd do that separately but I'm just wondering since adding a key value pair to the log entry is somewhat orthogonal to the logging level.

rainest replied:
https://github.com/Kong/kubernetes-ingress-controller/commit/572c2f448d53db90e3c4a9c988a6a965168d0b88 is the mechanism.

Levels are now verbosity values by convention, and controller-runtime has no less verbose level than what's effectively info. It has some things that are warning-ish at info, and some others that mention it in the message body: https://gist.github.com/rainest/e55fe44e54e4e1ab55d56f1210afb144#file-controller_runtime_logs-txt

I don't think there's any way to give controller-runtime a logger that bumps verbosity by one if we wanted to use a 0=warning, 1=info, 2=debug convention for our own logs.

On review of what we [were logging at warning](https://gist.github.com/rainest/e55fe44e54e4e1ab55d56f1210afb144#file-existing_kic_warnings-txt), it looks like they're all reasonable to just bump to error (none of them are things that should happen), so e1fd0608af7228fb62927cb0ee9815b1f8766fec

pmalek replied:
Given that all warning logs where converted to error now I'm 👍 this change.

If it ever comes to a point where we need warning logs, we can solve that then.

randmonkey reviewed (commented):
unit tests fails with the following failures:
```
    failures_test.go:103: 
        	Error Trace:	/home/runner/work/kubernetes-ingress-controller/kubernetes-ingress-controller/internal/dataplane/failures/failures_test.go:103
        	Error:      	Not equal: 
        	            	expected: 0
        	            	actual  : 2
        	Test:       	TestResourceFailuresCollector/does_not_crash_but_logs_warning_when_no_causing_objects_passed
```
Please take a look at the related code or UT.

pmalek commented on internal/store/store.go:
```diff
@@ -512,7 +513,7 @@ func (s Store) ListIngressClassParametersV1Alpha1() []*kongv1alpha1.IngressClass
 		classParam, ok := item.(*kongv1alpha1.IngressClassParameters)
 		if !ok {
-			s.logger.Warnf("listIngressClassParametersV1alpha1: dropping object of unexpected type: %#v", item)
+			s.logger.Error(nil, "listIngressClassParametersV1alpha1: dropping object of unexpected type", "type", fmt.Sprintf("%#v", item))
```
```suggestion
			s.logger.Error(nil, "listIngressClassParametersV1alpha1: dropping object of unexpected type", "type", fmt.Sprintf("%T", item))
```

pmalek replied:
I believe this doesn't happen too often but I'd change all those calls to what's in the suggestion as that's the actual "type" of the objects, additionally we could add a field with the `%#v` of the `item`.

I can see it's done similarly in other log site so it might be done in a separate PR to prevent this one from getting even bigger.

pmalek commented on internal/store/store.go:
```diff
@@ -490,7 +491,7 @@ func (s Store) ListIngressClassesV1() []*netv1.IngressClass {
 		class, ok := item.(*netv1.IngressClass)
 		if !ok {
-			s.logger.Warnf("listIngressClassesV1: dropping object of unexpected type: %#v", item)
+			s.logger.Error(nil, "listIngressClassesV1: dropping object of unexpected type", "type", fmt.Sprintf("%#v", item))
```
This is done in several places here:

What sort of worries me is using the error log with `error` set to `nil`. I'd much rather prefer to see a concrete error value, reused throughout the code rather than passing `nil` here and trying to be descriptive in the message.

pmalek replied:
Would this make us log

```
{"level":"error","time":"2023-09-22T00:28:13.006Z","error":"nil", "msg":"..."}
```

?

If that's the case then it'd be somewhat fishy for me if I saw something like this in my logs.

rainest replied:
It's omitted:

```
diff --git a/internal/manager/run.go b/internal/manager/run.go
index 48ac1ab0..44d8b458 100644
--- a/internal/manager/run.go
+++ b/internal/manager/run.go
@@ -52,6 +52,7 @@ func Run(
        logger logr.Logger,
 ) error {
        setupLog := ctrl.LoggerFrom(ctx).WithName("setup")
+       setupLog.Error(nil, "i am an error")
        setupLog.Info("starting controller manager", "release", metadata.Release, "repo", metadata.Repo, "commit", metadata.Commit)
        setupLog.Info("the ingress class name has been set", "value", c.IngressClassName)
```

```
2023-10-12T09:54:42-07:00	error	setup	i am an error

{"level":"error","time":"2023-10-12T09:56:42-07:00","logger":"setup","msg":"i am an error"}
```

pmalek reviewed (commented):
Just some nits and the conflicts (once again) and I believe this is ready  👍

pmalek commented:
Might want to adjust the PR title to e.g. adhere to conventional commits.

rainest commented:
> Might want to adjust the PR title to e.g. adhere to conventional commits.

I override the squash commit title and message if I use it instead of rebase and merge.
